engines/sherlock/screen.cpp

```diff
@@ -0,0 +1,528 @@
+	_fontNumber = fontNumb;
+	Common::String fname = Common::String::format("FONT%d.VGS", fontNumb + 1);
+
+	// Discard any previous font and read in new one
+	delete _font;
+	_font = new ImageFile(fname);
+
+	// Iterate through the frames to find the tallest font character
+	_fontHeight = 0;
+	for (uint idx = 0; idx < _font->size(); ++idx)
+		_fontHeight = MAX((uint16)_fontHeight, (*_font)[idx]._frame.h);
+}
+
+/**
+ * Handles updating any dirty areas of the screen Surface object to the physical screen
+ */
+void Screen::update() {
+	// Merge the dirty rects
+	mergeDirtyRects();
+
+	// Loop through copying dirty areas to the physical screen
+	Common::List<Common::Rect>::iterator i;
+	for (i = _dirtyRects.begin(); i != _dirtyRects.end(); ++i) {
+		const Common::Rect &r = *i;
+		const byte *srcP = (const byte *)getBasePtr(r.left, r.top);
+		g_system->copyRectToScreen(srcP, this->pitch, r.left, r.top,
+			r.width(), r.height());
+	}
+
+	// Signal the physical screen to update
+	g_system->updateScreen();
+	_dirtyRects.clear();
+}
+
+/**
+ * Return the currently active palette
+ */
+void Screen::getPalette(byte palette[PALETTE_SIZE]) {
+	g_system->getPaletteManager()->grabPalette(palette, 0, PALETTE_COUNT);
+}
+
+/**
+ * Set the palette
+ */
+void Screen::setPalette(const byte palette[PALETTE_SIZE]) {
+	g_system->getPaletteManager()->setPalette(palette, 0, PALETTE_COUNT);
+}
+
+/**
+ * Fades from the currently active palette to the passed palette
+ */
+int Screen::equalizePalette(const byte palette[PALETTE_SIZE]) {
+	int total = 0;
+	byte tempPalette[PALETTE_SIZE];
+	getPalette(tempPalette);
+
+	// For any palette component that doesn't already match the given destination
+	// palette, change by 1 towards the reference palette component
+	for (int idx = 0; idx < PALETTE_SIZE; ++idx) {
+		if (tempPalette[idx] > palette[idx])
+		{
+			tempPalette[idx] = MAX((int)palette[idx], (int)tempPalette[idx] - 4);
+			++total;
+		} else if (tempPalette[idx] < palette[idx]) {
+			tempPalette[idx] = MIN((int)palette[idx], (int)tempPalette[idx] + 4);
+			++total;
+		}
+	}
+
+	if (total > 0)
+		// Palette changed, so reload it
+		setPalette(tempPalette);
+
+	return total;
+}
+
+/**
+ * Fade out the palette to black
+ */
+void Screen::fadeToBlack(int speed) {
+	byte tempPalette[PALETTE_SIZE];
+	Common::fill(&tempPalette[0], &tempPalette[PALETTE_SIZE], 0);
+
+	while (equalizePalette(tempPalette)) {
+		_vm->_events->delay(15 * speed);
+	}
+
+	setPalette(tempPalette);
+}
+
+/**
+ * Fade in a given palette
+ */
+void Screen::fadeIn(const byte palette[PALETTE_SIZE], int speed) {
+	int count = 50;
+	while (equalizePalette(palette) && --count) {
+		_vm->_events->delay(15 * speed);
+	}
+
+	setPalette(palette);
+}
+
+/**
+ * Adds a rectangle to the list of modified areas of the screen during the
+ * current frame
+ */
+void Screen::addDirtyRect(const Common::Rect &r) {
+	_dirtyRects.push_back(r);
+	assert(r.isValidRect() && r.width() > 0 && r.height() > 0);
```

Comment:
Maybe not so much now, but it came in handy during development. And since we've still got another game to add support for, I'd prefer to leave it in place.